File: go-RayTracer/material.go
package main

import (
	"math"
	"math/rand"
)

// Lambertian is a diffuse material
type Lambertian struct {
	Albedo Color
}

// Scatter calculates how light scatters from a Lambertian surface
func (l *Lambertian) Scatter(rIn Ray, rec HitRecord) (bool, Color, Ray) {
	scatterDirection := rec.Normal.Add(RandomUnitVector())

	// Catch degenerate scatter direction
	if scatterDirection.NearZero() {
		scatterDirection = rec.Normal
	}

	scattered := Ray{rec.P, scatterDirection}
	attenuation := l.Albedo
	return true, attenuation, scattered
}

// Metal is a reflective material
type Metal struct {
	Albedo Color
	Fuzz   float64 // 0 for perfect mirror, higher for fuzzier reflection
}

// Scatter calculates how light scatters from a metal surface
func (m *Metal) Scatter(rIn Ray, rec HitRecord) (bool, Color, Ray) {
	reflected := Reflect(UnitVector(rIn.Direction()), rec.Normal)

	// Add fuzziness to the reflection
	fuzz := math.Min(m.Fuzz, 1.0)
	fuzzVector := RandomUnitVector().MulScalar(fuzz)
	reflected = reflected.Add(fuzzVector)

	scattered := Ray{rec.P, reflected}
	attenuation := m.Albedo

	// Only scatter if the ray reflects outward
	return Dot(scattered.Direction(), rec.Normal) > 0, attenuation, scattered
}

// Dielectric is a transparent material like glass or water
type Dielectric struct {
	RefractionIndex float64
}

// Scatter calculates how light scatters through a dielectric surface
func (d *Dielectric) Scatter(rIn Ray, rec HitRecord) (bool, Color, Ray) {
	attenuation := Color{[3]float64{1.0, 1.0, 1.0}} // Glass doesn't absorb any light
	var refractionRatio float64
	if rec.FrontFace {
		refractionRatio = 1.0 / d.RefractionIndex
	} else {
		refractionRatio = d.RefractionIndex
	}

	unitDirection := UnitVector(rIn.Direction())
	cosTheta := math.Min(Dot(unitDirection.Neg(), rec.Normal), 1.0)
	sinTheta := math.Sqrt(1.0 - cosTheta*cosTheta)

	// Determine if we must reflect
	cannotRefract := refractionRatio*sinTheta > 1.0
	var direction Vec3

	if cannotRefract || d.reflectance(cosTheta, refractionRatio) > rand.Float64() {
		direction = Reflect(unitDirection, rec.Normal)
	} else {
		direction = Refract(unitDirection, rec.Normal, refractionRatio)
	}

	scattered := Ray{rec.P, direction}
	return true, attenuation, scattered
}

// reflectance calculates the Schlick approximation
func (d *Dielectric) reflectance(cosine, refIdx float64) float64 {
	// Use Schlick's approximation for reflectance
	r0 := (1.0 - refIdx) / (1.0 + refIdx)
	r0 = r0 * r0
	return r0 + (1.0-r0)*math.Pow((1.0-cosine), 5)
}

File: ts-RayTracer/src/material.ts
import { Ray } from "./ray";
import { HitRecord, Material } from "./hittable";
import { Color, Vec3, reflect, refract, unitVector, randomUnitVector } from "./vec3";

export class Lambertian implements Material {
  constructor(public albedo: Color) {}
  scatter(rIn: Ray, rec: HitRecord) {
    let scatterDirection = rec.normal.add(randomUnitVector());
    if (scatterDirection.nearZero()) scatterDirection = rec.normal;
    const scattered = new Ray(rec.p, scatterDirection);
    const attenuation = this.albedo;
    return { didScatter: true, attenuation, scattered };
  }
}

export class Metal implements Material {
  constructor(public albedo: Color, public fuzz: number) {}
  scatter(rIn: Ray, rec: HitRecord) {
    const reflected = reflect(unitVector(rIn.direction()), rec.normal);
    const fuzzClamped = Math.min(this.fuzz, 1.0);
    const fuzzVector = randomUnitVector().mulScalar(fuzzClamped);
    const scattered = new Ray(rec.p, reflected.add(fuzzVector));
    const attenuation = this.albedo;
    const didScatter = dot(scattered.direction(), rec.normal) > 0;
    return { didScatter, attenuation, scattered };
  }
}

export class Dielectric implements Material {
  constructor(public refractionIndex: number) {}
  scatter(rIn: Ray, rec: HitRecord) {
    const attenuation = new Vec3([1.0, 1.0, 1.0]);
    const refractionRatio = rec.frontFace ? 1.0 / this.refractionIndex : this.refractionIndex;
    const unitDirection = unitVector(rIn.direction());
    const cosTheta = Math.min(dot(unitDirection.neg(), rec.normal), 1.0);
    const sinTheta = Math.sqrt(1.0 - cosTheta * cosTheta);
    const cannotRefract = refractionRatio * sinTheta > 1.0;
    let direction: Vec3;
    if (cannotRefract || this.reflectance(cosTheta, refractionRatio) > Math.random()) {
      direction = reflect(unitDirection, rec.normal);
    } else {
      direction = refract(unitDirection, rec.normal, refractionRatio);
    }
    const scattered = new Ray(rec.p, direction);
    return { didScatter: true, attenuation, scattered };
  }
  private reflectance(cosine: number, refIdx: number): number {
    let r0 = (1.0 - refIdx) / (1.0 + refIdx);
    r0 = r0 * r0;
    return r0 + (1.0 - r0) * Math.pow(1.0 - cosine, 5);
  }
}

// local util to avoid cycle
function dot(u: Vec3, v: Vec3): number {
  return u.e[0] * v.e[0] + u.e[1] * v.e[1] + u.e[2] * v.e[2];
}


